staging/src/k8s.io/apiserver/pkg/server/options/etcd.go

```diff
@@ -213,12 +218,42 @@ func (s *EtcdOptions) Complete(storageObjectCountTracker flowcontrolrequest.Stor
 	}
 
 	if len(s.EncryptionProviderConfigFilepath) != 0 {
-		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, stopCh)
+		ctx, closeTransformers := wait.ContextForChannel(stopCh)
+		transformerOverrides, kmsPluginHealthzChecks, err := encryptionconfig.LoadEncryptionConfig(s.EncryptionProviderConfigFilepath, ctx.Done())
 		if err != nil {
+			// in case of error, we want to close partially initialized (if any) transformers
+			closeTransformers()
 			return err
 		}
-		s.transformerOverrides = transformerOverrides
+
 		s.kmsPluginHealthzChecks = kmsPluginHealthzChecks
+		s.dynamicTransformers = encryptionconfig.NewDynamicTransformers(transformerOverrides, closeTransformers, stopCh)
+
+		// add post start hook to start hot reload controller
+		// adding this hook here will ensure that it gets configured exactly once
+		err = addPostStartHook(
+			"start-kms-encryption-config-hot-reloader",
+			func(hookContext server.PostStartHookContext) error {
+				// get context
+				// we ignore the cancel func because this context should only be canceled when stopCh is closed
+				ctx, _ := wait.ContextForChannel(hookContext.StopCh)
+
+				kmsConfigController := kmsconfigcontroller.NewDynamicKMSEncryptionConfiguration(
+					"kms-encryption-config",
+					s.EncryptionProviderConfigFilepath,
+					s.dynamicTransformers,
+					hookContext.StopCh,
```

Comment:
All `LoadEncryptionConfig` calls should use the same channel.

```suggestion
					stopCh,
```